enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+---
+title: ingress-node-firewall-with-ebpf
+
+authors:
+  - "@msherif1234"
+
+reviewers:
+  - "@danwinship"
+  - "@dave-tucker"
+  - "@dcbw"
+  - "@trozet"
+
+approvers:
+  - "@danwinship"
+  - "@knobunc"
+
+api-approvers:
+  - "@danwinship"
+
+creation-date: 2022-06-15
+
+last-updated: 2022-06-21
+
+status: implementable
+
+tracking-link:
+  - https://issues.redhat.com/browse/SDN-2781
+---
+
+# Ingress node Firewall using eBPF
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OCP customers would like to secure OCP's nodes from external attacks by configuring ingress firewall policies.
+eBPF provides a very flexible filtering mechansim to allow early detection and filtering.
+eXpress Data Path (XDP) is a feature in the Linux kernel which allows users to execute a user-supplied eBPF program when a packet is received on a network interface (NIC). 
+This program is executed in the NIC’s device driver, right after the interrupt processing and before the network stack. 
+The goal of XDP is to offer comparable performance to kernel bypass solutions while working with the existing kernel networking stack.
+New cluster scoped object `IngressNodeFirewall` will be used to configure node ingress firewall match rules and a per matching rule statistics will be avaliable.
+Admin can provision multiple instances of this object while ensuring certain execution order.
+Another cluster scoped object `NodeEndpoint` which will be used to specify the attachment interface(s) for example admin can apply specific node firewall rules on one or more interfaces and another different set of rules on another interface(s) or apply the same rules to all interfaces.
+For the best performance XDP programs need to attach to interfaces using the following models:
+- Native XDP
+  The XDP program is loaded by the network card driver as part of its initial receive path. This also requires support from the network card driver.
+- Offloaded XDP
+  The XDP program loads directly on the NIC, and executes without using the CPU. This requires support from the network interface device.
+
+## Motivation
+
+Customers would like to secure OCP cluster from external attacks.
```

Comment:
I think this needs to be narrowed down... If the ask is this wide, I simple "node firewall" is too narrow of a solution.  

This motivation makes me think that users really wan't to apply policy to all possible means of ingressing a cluster.  Such an object would need to be implemented across the CNI and Ingress engine to be feasible. 

I just don't want us to spend the time making an OCP specific object for customers just for them to eventually come back and say "we need more"

Reply:
@astoycos this is already solved prob with other vendors both Calico and Cillium have similar approach for node protection and this is just a gap for OCP , so while we are trying to fill in this gap we trying to make the api flexibile as much as we can

Reply:
Right but in Cilium's case they simply extend their cluster wide network policy via a nodelabel selector https://docs.cilium.io/en/v1.9/gettingstarted/host-firewall/ 

And Calico does something similar https://projectcalico.docs.tigera.io/security/protect-hosts 

Really what i'm trying to say, it doesn't seem right to have a dedicated API object with a single use case ... i.e node firewalling... 

Since once we let ANY API object out of the box we will be stuck with it for many many releases 

Why not look at following the same schema used by cilium and Calico... I.e We have a cluster wide policy object in the new [AdminNetworkPolicy API](https://github.com/kubernetes-sigs/network-policy-api) And we could work together with upstream so simply add N/S traffic support + a nodeselector  

Sorry If I'm being pedantic here @msherif1234 I just think API design is one of the most important parts here, and IMO the more APIs we can share with the upstream community the easier it will be for CU's and users in general

Reply:
we did look at both implementation, also it was mentioned ANP earlier we decided to decouple them at least for this release specially that early phases on ANP is just for intercluster traffic unlike what we need to do here when that matures we can see possible leverage in the future, I agree API design is very critical piece that is why we put too much effort defining it, also this downstream only project